src/clp_ffi_py/ir/native/PySerializer.cpp

```diff
@@ -0,0 +1,569 @@
+#include <clp_ffi_py/Python.hpp>  // Must always be included before any other header files
+
+#include "PySerializer.hpp"
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <optional>
+#include <span>
+#include <type_traits>
+#include <utility>
+
+#include <clp/ffi/ir_stream/protocol_constants.hpp>
+
+#include <clp_ffi_py/api_decoration.hpp>
+#include <clp_ffi_py/error_messages.hpp>
+#include <clp_ffi_py/ir/native/error_messages.hpp>
+#include <clp_ffi_py/PyObjectCast.hpp>
+#include <clp_ffi_py/PyObjectUtils.hpp>
+#include <clp_ffi_py/utils.hpp>
+
+namespace clp_ffi_py::ir::native {
+namespace {
+/**
+ * Callback of `PySerializer`'s `__init__` method:
+ */
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
+PyDoc_STRVAR(
+        cPySerializerDoc,
+        "Serializer for CLP key-value pair IR stream.\n"
```

Comment:
For all these docstrings you can use...

```
:type <param>: <type>
:rtype: <type>
```

... to document the parameter types and return types.

E.g.:
```py
"""
:type msgpack_map: bytes\n
:rtype: int\n
"""
```

Reply:
Done for all user-level APIs. Didn't add for `__exit__` and `__enter__` since they won't be rendered by the doc site. And users aren't supposed to call these methods explicitly anyways